Chatkit/Auth/TokenProvider.swift

```diff
@@ -0,0 +1,137 @@
+    /// - Returns:a `TokenProvider` which fetches tokens from the Chatkit Test Token Provider.
+    public static func chatkitTestTokenProvider(instanceLocator: String, userId: String) -> TokenProvider {
+        return CachingTokenProvider(
+            delegateProvider: RetryingTokenProvider(
+                delegateProvider: testTokenHTTPCallProvider(
+                    instanceLocator: instanceLocator, userId: userId
+                ),
+                retryAttempts: 3,
+                retryDelayMs: 1000
+            )
+        )
+    }
+
+    /// Create a `TokenProvider` which fetches tokens from an HTTP endpoint in your backend, with caching and retry implemented. This is the recommended implementation for production use.
+    ///
+    /// - Parameters:
+    ///     - method: The HTTP method to use in the call, e.g. `GET`, `POST`
+    ///     - host: The host name to be called, e.g. `example.com`
+    ///     - path: The path component of the URL to be called, e.g. `/tokens`
+    ///     - headers: An optional map of headers to include in the request. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+    ///     - queryParams: An optional map of query parameters to include in the request URL. Here you can supply the session or other credientials which your endpoint might require to authenticate the request.
+    /// - Returns: a `TokenProvider` which fetches tokens from the specified HTTP endpoint, and is wrapped in default caching and retry logic.
+    public static func httpTokenProvider(method: String, host: String, path: String, headers: [String: String]?, queryParams: [String: String]?) -> TokenProvider {
+        return CachingTokenProvider(
+            delegateProvider: RetryingTokenProvider(
+                delegateProvider: HTTPCallTokenProvider(
+                    method: method,
+                    host: host,
+                    path: path,
+                    headers: headers,
+                    queryParams: queryParams
+                ),
+                retryAttempts: 3,
+                retryDelayMs: 1000
+            )
+        )
+    }
+
+    internal static func testTokenHTTPCallProvider(instanceLocator: String, userId: String) -> TokenProvider {
+        // TODO: Implement:
+        // extract host and instanceId from instanceLocator
+        let host = "us1.pusherplatform.io"
+        let instanceId = "UNIMPLEMENTED"
+        let path = "/services/chatkit_token_provider/v1/\(instanceId)/token"
+
+        return HTTPCallTokenProvider(
+            method: "POST",
+            host: host,
+            path: path,
+            headers: nil,
+            queryParams: [ "user_id": userId ]
+        )
+    }
+}
+
+/// CachingTokenProvider delegates the fetching of tokens to an underlying `delegateProvider`, retrying on failure
+public class CachingTokenProvider: TokenProvider {
```

Comment:
Perhaps it would be worth moving `CachingTokenProvider`, `RetryingTokenProvider` and `HTTPCallTokenProvider` classes to separate files?

Reply:
Sure, if we decide to keep them. I don't know if there's a better way to introduce the desired common functionality?